Feature/module arrange

lgannoaa opened the pull request:
Provide solution to issue #479 - Review modules being loaded on each platform (e.g. via module_base.*) to check they are loading in the proper order and dependencies are being met correctly. Check for module swaps and resolve as needed to maintain the correct module loads.
Change made to:
   OznMonBuild.orion, RadMonBuild.orion - Fix issue with resolving hpc-intel version. 
   module_base.orion - Fix issue with module swaps and arrange module for proper order.
   modulefile.ww3.orion - Fix issue with module swaps.
   module_base.wcoss_dell_p3 - Fix issue with missing version number in pm5 load. Fix issue with module swaps and arrange module for proper order.
   module_base.hera -  Fix issue with module swaps.
   modulefile.ww3.hera - Fix issue with module swaps.

lgannoaa committed:
Fixed global workflow develop to build and link on JET

lgannoaa committed:
This version of the ecflow is used to run low resolution cycled parallel on global-workflow develop.
Support C192C96L127 on WCOSS p3
Step-By-Step documentation is located on Google DOC:
  GWD EMC ECFLOW LOWRES HOWTO
  https://docs.google.com/presentation/d/1ImLJpFaw9WJj-ghpzqcs727xdXHeQMhMkeK7TSqwocA/edit?usp=sharing
Start up configuration script is in $HOMEgfs/ecflow/ecflow_config_wcoss_p3_c192c96l127.sh
Modify the ECFLOW_ROOT, PTMP. DATAROOT, and ECF_HOST for your parallel information.
Run this script and follow the on-line document to startup the parallel.

lgannoaa committed:
Merge with JET develop fcst only change from GW develop

lgannoaa committed:
Merge remote-tracking branch 'upstream/develop' into develop

lgannoaa committed:
sync merge

lgannoaa committed:
Remove gfs_util.jet

lgannoaa committed:
Merge remote-tracking branch 'upstream/develop' into develop

lgannoaa committed:
Update ORION module files - Module order and dependency check (#479)
  RadMonBuild.orion - Update and rearrange module. Remove unnecessary environment variable export
  OznMonBuild.orion - Update and rearrange module. Remove unnecessary environment variable export
  modulefile.ww3.orion - Update module version.
  module_base.orion - Update and rearrange module.

lgannoaa committed:
Update HERA module files - Module order and dependency check (#479)
  modulefiles/module_base.hera - Update and rearrange module.
  modulefiles/modulefile.ww3.hera - Update and rearrange module.

lgannoaa committed:
Update wcoss_dell_p3 module files - Module order and dependency check (#479)
  modulefiles/module_base.wcoss_dell_p3
    - Update and rearrange module.
    - Implemented workaround for hpc-stack bug in github #370

lgannoaa committed:
Cleanup modulefile OznMonBuild and RadMonBuild for orion

lgannoaa committed:
Merge branch 'feature/module-arrange' of https://github.com/lgannoaa/global-workflow into feature/module-arrange

lgannoaa commented:
@KateFriedman-NOAA let me know if this draft PR look good to you. I will make it become a regular PR.

aerorahul commented on modulefiles/module_base.orion:
```diff
@@ -1,52 +1,41 @@
 module load w3nco/2.4.1
-module load upp/10.0.6
-
+module load jasper/2.0.25
+module load zlib/1.2.11
+module load png/1.6.35
```
please move these back to their original location.  These are prerequisites to some other libraries that are loaded before.

lgannoaa replied:
upp need to be load at the time g2 and g2tmpl is expected to be loaded. Because loading upp will also load g2 and g2tmpl automatically.

aerorahul replied:
This comment was for jasper, zlib and png.  How does upp figure in the response to this comment?

lgannoaa replied:
Answered in Files changed page.

aerorahul commented on modulefiles/module_base.orion:
```diff
@@ -1,52 +1,41 @@
+# waveprep
+module load cdo/1.9.5
 
-module load hdf5/1.10.6
-module load netcdf/4.7.4
```
why are these removed?

lgannoaa replied:
Answered in Files changed page.

aerorahul commented on modulefiles/module_base.orion:
```diff
@@ -1,52 +1,41 @@
-module load zlib/1.2.11
-module load png/1.6.35
+# waveprep
+module load cdo/1.9.5
```
This module can only be loaded after hdf5 and netcdf are loaded.

Please do not move the module loads around.
The order of module loads is critical.

lgannoaa replied:
Orion-login-1[1] /home/lgan$ module purge
Orion-login-1[2] /home/lgan$ module use /apps/contrib/NCEP/libs/hpc-stack/modulefiles/stack
Orion-login-1[3] /home/lgan$ module load hpc/1.1.0
Orion-login-1[4] /home/lgan$ module load hpc-intel/2018.4
Orion-login-1[5] /home/lgan$ module unload mkl/2020.2
Orion-login-1[6] /home/lgan$ module load mkl/2018.4
Orion-login-1[7] /home/lgan$
Orion-login-1[7] /home/lgan$ module load hpc-impi/2018.4
Orion-login-1[8] /home/lgan$ module load slurm/19.05.3-2
Orion-login-1[9] /home/lgan$ module load cdo/1.9.5
Orion-login-1[10] /home/lgan$ module list

Currently Loaded Modules:
  1) hpc/1.1.0   2) intel/2018.4   3) hpc-intel/2018.4   4) mkl/2018.4   5) impi/2018.4   6) hpc-impi/2018.4   7) munge/0.5.13   8) slurm/19.05.3-2   9) hdf5/1.10.6  10) netcdf/4.7.4  11) cdo/1.9.5

This shows that the hdf5 and netcdf is loaded after cdo module.

lgannoaa replied:
Answered in Files changed page.

lgannoaa replied:
In a global-workflow PR's meeting on 12/8 3:45p, Rahul requested to module load netcdf and hdf5 before module load cdo. Keep module load cdo on the bottom of this file.

aerorahul commented on modulefiles/module_base.orion:
```diff
@@ -1,52 +1,41 @@
 module load bacio/2.4.1
-module load g2/3.4.1
-module load g2tmpl/1.9.1
+module load upp/10.0.6
```
UPP module requires several of the libraries loaded.
Please don't move the module loads around unless you have a sound reason to do so.

lgannoaa replied:
load upp module also load g2 and g2tmpl automatically. That is the reason for the change.

lgannoaa replied:
Answered in Files changed page.

lgannoaa replied:
In a global-workflow PR's meeting on 12/8 3:45p, Rahul requested to keep module load upp in the same location, keep module load g2 and g2tmpl in the same location, and update the version number for g2 and g2tmpl to match upp.

aerorahul commented on modulefiles/module_base.wcoss_dell_p3:
```diff
@@ -61,3 +45,13 @@ module load python/3.6.3
+
+# Workaround for github hpc-stack issues #370
+setenv LMOD_EXACT_MATCH no
+module load netcdf/4.7.4
+module load nemsio/2.5.2
+module load w3emc/2.7.3
+module load upp/10.0.6
+module load wgrib2/2.0.8
+setenv WGRIB2 wgrib2
```
Please move this back to where they were in the module load.

lgannoaa replied:
These module need not to be load with LMOD_EXACT_MATCH=no. Therefore they should be load before the LMOD_EXACT_MATCH=no statement.

lgannoaa replied:
Answered in Files changed page.

aerorahul commented on modulefiles/module_base.wcoss_dell_p3:
```diff
@@ -61,3 +45,13 @@ module load python/3.6.3
 module load cdo/1.9.8
+
+# Workaround for github hpc-stack issues #370
+setenv LMOD_EXACT_MATCH no
```
please set this variable immediately after `EnvVars` module is loaded.

lgannoaa replied:
LMOD_EXACT_MATCH=no effectively turn off all version check. It is not recommended practice.
The reason of setting LMOD_EXACT_MATCH=no is to provide a temporary workaround to address a known bug in wcoss_dell_p3 hpc-stack. Therefore, LMOD_EXACT_MATCH=no should only be used in the module impacted by this bug.

aerorahul replied:
`LMOD_EXACT_MATCH=no`, does not turn OFF version check.  It enables loading a `default` version designated, if a version is not specified.  It has no impact if a module is already loaded.

WCOSS2 does not require this any more.

lgannoaa replied:
Answered in Files changed page.

lgannoaa replied:
@aerorahul 
This file module_base.wcoss_dell_p3 is not for wcoss2. 
Do you want me to set LMOD_EXACT_MATCH=no for the entire module_base.wcoss_dell_p3 file? 
Do you want me to not set LMOD_EXACT_MATCH=yes at the end of module_base.wcoss_dell_p3 file?

lgannoaa replied:
In a global-workflow PR's meeting on 12/8 3:45p, Rahul requested to set LMOD_EXACT_MATCH=no for the entire module_base.wcoss_dell_p3.

aerorahul commented on modulefiles/module_base.wcoss_dell_p3:
```diff
@@ -61,3 +45,13 @@ module load python/3.6.3
+module load upp/10.0.6
+module load wgrib2/2.0.8
+setenv WGRIB2 wgrib2
+setenv LMOD_EXACT_MATCH yes
```
This likely has no impact.

lgannoaa replied:
LMOD_EXACT_MATCH need to be turn on as default. That is the recommended best practice. Therefore, it is turned on after the temporary workaround solution has applied.

lgannoaa replied:
Answered in Files changed page.

aerorahul reviewed (changes requested):
Comments:
- versions updates will need to be tested. this is ok.
- the order in which modules are loaded is critical for ensuring a robust and reproducible software stack.  the order ensures prerequisites are available for subsequent modules.  please revert the order in which the modules are loaded, unless there was a bug encountered.

lgannoaa committed:
Module order and dependency check (#479)
In a global-workflow PR's meeting on 12/8 3:45p, Rahul requested:
 - set LMOD_EXACT_MATCH=no for the entire module_base.wcoss_dell_p3 after loading EnvVars module.
 - module load upp in the current location, keep module load g2 and g2tmpl in the same location, and
 update the version number for g2 and g2tmpl to match upp. Same approach apply to cdo module.
modified:   module_base.orion
modified:   module_base.wcoss_dell_p3

lgannoaa commented:
@aerorahul, I updated the module file as you requested yesterday. Please take a look.

aerorahul commented on modulefiles/module_base.orion:
```diff
@@ -1,15 +1,16 @@
 module load hpc/1.1.0
 module load hpc-intel/2018.4
+
+# gsi
+module unload mkl/2020.2
+module load mkl/2018.4
```
why is this necessary here?  The GSI modulefile at build time does this switch.
Can you explain why/where this is used at runtime?

lgannoaa replied:
Will contact GSI team for guidance.

RussTreadon-NOAA replied:
Background information is provided in NOAA-EMC/GSI issue [#110](https://github.com/NOAA-EMC/GSI/issues/110) and NOAA-EMC/hpc-stack issue [#170](https://github.com/NOAA-EMC/hpc-stack/issues/170).   @MichaelLueken-NOAA added to keep him in the loop.

aerorahul commented on modulefiles/module_base.orion:
```diff
@@ -1,15 +1,16 @@
+module load mkl/2018.4
 module load hpc-impi/2018.4
 
-#module load hpss/hpss
```
I can understand there is no `hpss` module.
Is `nco` not used on Orion?

aerorahul commented on modulefiles/module_base.wcoss_dell_p3:
```diff
@@ -49,7 +52,6 @@ module load upp/10.0.6
 module load wgrib2/2.0.8
 setenv WGRIB2 wgrib2
 
-module use -a /gpfs/dell1/nco/ops/nwprod/modulefiles/
```
are we sure we don't need this line?  How will `gempak` module be found? Is it in the default `MODULEPATH`?

aerorahul reviewed (changes requested):
This PR is now reduced to the following items:

1. versions for a few libraries `g2`, `g2tmpl`, `w3emc` have been updated across `hera/orion/wcoss_dell_p3`.  Has this been robustly tested?  `wcoss_dell_p3` still uses a different `w3emc` library than on `Hera`?  What is the reason for this?
2. The Ozone and Radiance Monitoring build modulefiles should be deleted from this repository.  The GSI repository contains the relevant ones.  The ones in the global-workflow are obsolete.

There is no more "rearrangement" of the modules here as the PR title and description suggests.

KateFriedman-NOAA commented:
This should probably be closed now. A number of the changed files in this draft PR have since been removed (e.g. OznMon/Radmon) or will be massively updated/renamed during the WCOSS2 port (LUA + hpc-stack updates). The module order is being addressed in the WCOSS2 port updates as well so this is a duplicate now. Thanks!
